server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -1389,16 +1409,23 @@ public String getProfileName() {
         @Override
         public void sendResponse(TransportResponse response) throws IOException {
             service.onResponseSent(requestId, action, response);
-            final TransportResponseHandler<?> handler = service.responseHandlers.onResponseReceived(requestId, service);
-            // ignore if its null, the service logs it
-            if (handler != null) {
+            try (var shutdownBlock = service.pendingDirectHandlers.withRef()) {
```

Comment:
I'm struggling with how this would work in case of a handler running on the generic pool. We increment the pending handlers here, then enqueue the task on the generic queue and then decrement.
Wouldn't we still complete the stop of the transport service now before that task has been executed and potentially just quietly kill the generic pool without ever running our `ForkingResponseHandlerRunnable`? If it's enqueued and `shutdown` is called on that queue, I don't think there's anything that notifies our task of the fact that it will never run?
Wouldn't we have to push the decrementing of the ref on `shutdownBlock` to the generic pool in this example to actually be 100% safe?

Reply:
My thinking is that it's enough to enqueue everything, we don't have to wait for them all to execute, because we start the shutdown by calling `ThreadPool#shutdown()` which should drain the threadpool queue first. It's true that this doesn't properly guarantee that everything executes (we call `shutdownNow()` after 10s regardless) but in practice this only matters for tests which should be cleaning everything up in good time anyway.

I suppose we could assert that `shutdownNow()` returns an empty list in most tests to be more explicit about that. Not sure I want to do that here tho.

Reply:
> I suppose we could assert that shutdownNow() returns an empty list in most tests to be more explicit about that. Not sure I want to do that here tho.

Right ... that's what we actually want in a follow-up. I'm 95% sure that will trip here and there ...

Reply:
Ah I checked and it looks like we almost have this already. We throw an `AssertionError` if graceful shutdown fails in `ESSingleNodeTestCase` already but apparently only an `IOException` in `ESIntegTestCase`. I think that would result in test failures already but for the avoidance of doubt I opened https://github.com/elastic/elasticsearch/pull/85238 to make this a proper assertion.